client/state/data-layer/wpcom/login-2fa/index.js

```
@@ -20,77 +20,83 @@ import {
+			method: 'POST',
+			headers: [
+				[ 'Content-Type', 'application/x-www-form-urlencoded' ],
+			],
+			withCredentials: true,
+			body: {
+				user_id: getTwoFactorUserId( store.getState() ),
+				auth_type: authType,
+				two_step_nonce: getTwoFactorAuthNonce( store.getState(), authType ),
+				remember_me: getRememberMe( store.getState() ),
+				two_step_push_token: getTwoFactorPushToken( store.getState() ),
+				client_id: config( 'wpcom_signup_id' ),
+				client_secret: config( 'wpcom_signup_key' ),
+			},
+		},
+		action
+	) );
 };
 
+const receivedTwoFactorPushNotificationApproved = ( { dispatch } ) =>
+	dispatch( { type: TWO_FACTOR_AUTHENTICATION_PUSH_POLL_COMPLETED } );
+
 /***
- * Polling the login API for the status of the push notification.
- * The polling will stop on success or when store's polling progress
- * state changes to `false`
+ * Receive error from the two factor push notification status http request
  *
- * @param {Object}   store  Global redux store
+ * @param {Object}	store  Global redux store
+ * @param {Object}	action dispathced action
+ * @param {Object}	error the error object
  */
-const doAppPushPolling = store => {
-	let retryCount = 0;
-	const retry = () => {
-		// if polling was stopped or not in progress - stop
-		if ( ! getTwoFactorPushPollInProgress( store.getState() ) ) {
-			return;
-		}
+const receivedTwoFactorPushNotificationError = ( store, action, next, error ) => {
+	const twoStepNonce = get( error, 'response.body.data.two_step_nonce' );
 
-		retryCount++;
-		setTimeout(
-			() => doAppPushRequest( store ).catch( retry ),
-			( POLL_APP_PUSH_INTERVAL_SECONDS + Math.floor( retryCount / 10 ) ) * 1000 // backoff lineary
-		);
-	};
+	if ( ! twoStepNonce ) {
+		store.dispatch( { type: TWO_FACTOR_AUTHENTICATION_PUSH_POLL_STOP } ); //
+		throw new Error( "Two step nonce wasn't present on the response from polling endpoint, unable to continue" );
```

While I have mixed feelings leaning towards liking this approach of failing fast, I know that @blowery has made some good points about keeping exceptions out of our code.